prov/gni/src/gnix_shmem.c

```diff
@@ -0,0 +1,182 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+#include "gnix_shmem.h"
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+
+#include "rdma/fi_errno.h"
+#include "gnix_util.h"
+
+int _gnix_shmem_create(
+	const char *path,
+	uint32_t size,
+	int (*init_func)(const char *path, uint32_t size, void *region),
+	struct gnix_shared_memory *region)
+{
+	int rc, fd, saved_errno, count;
+	void *buffer;
+
+	if (!region || !path) {
+		GNIX_WARN(FI_LOG_FABRIC, "bad params, ret=-FINVAL");
+		return -FI_EINVAL;
+	}
+
+	fd = open(path, O_CREAT | O_EXCL | O_RDWR, 0755);
+	if (fd >= 0) {
+		buffer = malloc(size);
+		if (!buffer) {
+			GNIX_INFO(FI_LOG_FABRIC,
+				"failed to allocate memory for shared memory segment");
+			return -FI_ENOMEM;
+		}
+
+		if (init_func) {
+			rc = init_func(path, size, buffer);
+			if (rc != 0) {
+				free(buffer);
+				return rc;
+			}
+		} else
+			memset(buffer, 0, size);
+
+		GNIX_INFO(FI_LOG_FABRIC,
+			"initializing shared memory segment, path=%s\n",
+			path);
+
+		count = write(fd, buffer, size);
+		if (count != size) {
+			GNIX_WARN(FI_LOG_FABRIC,
+				"failed to write to shared memory segment, "
+				"expected=%d actual=%d\n",
+				size, count);
+			saved_errno = errno;
+			close(fd);
+			unlink(path);
+			free(buffer);
+			errno = saved_errno;
+			return -saved_errno;
+		}
+		free(buffer);
+		fchmod(fd,
+			S_IRUSR | S_IWUSR |
+			S_IRGRP | S_IWGRP |
+			S_IROTH | S_IWOTH);
+	} else {
+
+		/* Unexpected error? */
+		if (errno != EEXIST) {
+			saved_errno = errno;
+			GNIX_WARN(FI_LOG_FABRIC,
+				"open() of shared memory segment failed, errno=%d\n",
+				saved_errno);
+
+			return -errno;
+		}
+
+		/*
+		 * The file was created by another process, but it might not yet
+		 * be initialized.  Wait until it is.
+		 */
+		struct stat stat_info;
+		int wait_count = 0;
+
+		do {
+			GNIX_INFO(FI_LOG_FABRIC,
+				"stat() of shared memory segment.");
+			rc = stat(path, &stat_info);
+			if (rc != 0) {
+				GNIX_DEBUG(FI_LOG_FABRIC,
+					"stat() of shared memory segment "
+					"failed, errno %d\n", errno);
+				return -errno;
+			}
+
+			if (stat_info.st_size < size) {
+				GNIX_INFO(FI_LOG_FABRIC,
+					"Shared memory segment is %d bytes, "
+					"waiting for it to be %d\n",
+					stat_info.st_size, size);
+				sleep(1);
+				wait_count++;
+			}
+		} while ((stat_info.st_size < size) &&
```

Comment:
It would be nice to have a tunable here for how long to wait for the file to be ready.  I'm thinking of KNL nodes with very slow startup of processes under slurm.  check with MPT group for issues in that area.